File: mainframe_migration/examples/generated_targets/dotnet-bigquery/MfRuntime.cs
// Native C# helpers preserve source storage, record bytes, aliases and arithmetic.
// Source statements execute in generated job functions, not in another language.
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Numerics;

public static class MfRuntime {
    public static Dictionary<string,object> obj(object x)=>x as Dictionary<string,object> ?? throw new ArgumentException("Expected JSON object");
    public static List<object> list(object x)=>x as List<object> ?? throw new ArgumentException("Expected JSON array");
    public static Dec num(object x){if(x is Dec n)return n;if(x is string s)return Dec.Parse(s);if(x is int||x is long)return Dec.Parse(x.ToString());throw new ArgumentException("Expected exact number");}
    public static int integer(object x)=>num(x).ToInt32();
    public static object math(string op,object a,object b)=>Dec.Math(op,num(a),num(b));
    public static bool truth(object x)=>x is bool b?b:x is Dec n?!n.N.IsZero:x is string s?s.Length>0:throw new ArgumentException("Unsupported condition value");
    public static string pad(string s,int n)=>s.Length>=n?s.Substring(0,n):s.PadRight(n,' ');
    public static byte[] encode(string text,string table){var data=new byte[text.Length];for(int i=0;i<text.Length;i++){int n=table.IndexOf(text[i]);if(n<0)throw new ArgumentException("Character is outside source code page");data[i]=(byte)n;}return data;}
    public static string decode(byte[] data,string table){var result=new char[data.Length];for(int i=0;i<data.Length;i++){if(data[i]>=table.Length)throw new ArgumentException("Byte is outside source code page");result[i]=table[data[i]];}return new string(result);}
    public static int bytesCompare(byte[] a,byte[] b){for(int i=0;i<Math.Min(a.Length,b.Length);i++){int c=a[i].CompareTo(b[i]);if(c!=0)return c;}return a.Length.CompareTo(b.Length);}
    public sealed class Ref {public Fields Owner;public string Name;public Ref(Fields owner,string name){Owner=owner;Name=name;}}
    public sealed class Fields {
        readonly Dictionary<string,Dictionary<string,object>> layout=new();readonly Dictionary<string,char[]> buffers=new();readonly Dictionary<string,Ref> aliases=new();
        // Store groups and individual fields in the same underlying characters.
        public Fields(List<object> definitions){foreach(var x in definitions){var f=obj(x);string name=(string)f["name"],root=(string)f["root"];layout.Add(name,f);int size=integer(f["offset"])+integer(f["length"]);if(!buffers.TryGetValue(root,out var old)||old.Length<size){var next=Enumerable.Repeat(' ',size).ToArray();if(old!=null)Array.Copy(old,next,old.Length);buffers[root]=next;}}
            foreach(var f in layout.Values)if((string)f["kind"]=="number")set((string)f["name"],Dec.Zero);foreach(var f in layout.Values)if(f.ContainsKey("value"))set((string)f["name"],f["value"]);}
        (char[],int) storage(string name){var f=layout[name];string root=(string)f["root"];int offset=integer(f["offset"]);if(aliases.TryGetValue(root,out var r)){var (data,start)=r.Owner.storage(r.Name);return(data,start+offset);}return(buffers[root],offset);}
        public int length(string name)=>integer(layout[name]["length"]);
        public string raw(string name){var(data,start)=storage(name);return new string(data,start,length(name));}
        public object get(string name){string text=raw(name);var f=layout[name];if((string)f["kind"]!="number")return text;if(text.Length==0||text.Any(c=>c<'0'||c>'9'))throw new ArgumentException("Nonnumeric source field "+name);int scale=integer(f["scale"]);if(scale>0)text=text.Insert(text.Length-scale,".");return Dec.Parse(text);}
        // Use the receiving picture's low-order digits; never replace this with banker's rounding.
        public void set(string name,object value){var f=layout[name];int n=length(name);string text;if((string)f["kind"]=="number"){var number=BigInteger.Abs(num(value).Truncated(integer(f["scale"])));text=(number%BigInteger.Pow(10,n)).ToString().PadLeft(n,'0');}else text=pad(value.ToString(),n);var(data,start)=storage(name);text.CopyTo(0,data,start,n);}
        public Ref reference(string name){storage(name);return new Ref(this,name);}
        public void bind(string name,Ref reference){var f=layout[name];if((string)f["root"]!=name||integer(f["offset"])!=0||length(name)!=reference.Owner.length(reference.Name))throw new ArgumentException("Source CALL linkage differs in width/root");aliases[name]=reference;}
    }
    public sealed class RecordFile {
        readonly FileStream stream;readonly string mode,table;readonly int length;bool closed;
        public RecordFile(string path,string mode,string table,int length){this.mode=mode;this.table=table;this.length=length;if(mode=="OUTPUT")Directory.CreateDirectory(Path.GetDirectoryName(path));if(mode!="INPUT"&&mode!="OUTPUT")throw new ArgumentException("Unsupported file mode");stream=new FileStream(path,mode=="INPUT"?FileMode.Open:FileMode.Create,mode=="INPUT"?FileAccess.Read:FileAccess.Write);}
        public string read(){if(mode!="INPUT")throw new IOException("File is not open for input");var raw=new byte[length];int got=0,n;while(got<length&&(n=stream.Read(raw,got,length-got))>0)got+=n;if(got==0)return null;if(got!=length)throw new IOException("Input ends with a partial record");return decode(raw,table);}
        public void write(string text){var raw=encode(text,table);if(mode!="OUTPUT"||raw.Length!=length)throw new IOException("Output record length/mode differs from source");stream.Write(raw,0,raw.Length);}
        public void close(){if(closed)return;stream.Dispose();closed=true;}
    }
    public sealed class Context : IDisposable {
        public readonly Dictionary<string,object> payload;public readonly Dictionary<string,int> events=new();public readonly MfDatabase db;
        readonly Dictionary<string,object> datasets,tables;readonly List<RecordFile> opened=new();readonly string caseRoot,workRoot,encoding;
        public Context(Dictionary<string,object> p){payload=p;datasets=obj(p["datasets"]);tables=obj(p["codepages"]);encoding=(string)p["collation"];caseRoot=Path.GetFullPath((string)p["case_root"]);workRoot=Path.GetFullPath((string)p["work_root"]);db=new MfDatabase(p);}
        public void hit(string id){events[id]=events.GetValueOrDefault(id,0)+1;}
        public List<object> layout(string program){foreach(var m in list(payload["models"])){var model=obj(m);if((string)model["job"]!=(string)payload["job"])continue;foreach(var p in list(model["programs"])){var spec=obj(p);if((string)spec["name"]==program)return list(spec["layout"]);}}throw new ArgumentException("Program layout is missing");}
        public Dictionary<string,object> dd(string job,string step){foreach(var m in list(payload["models"])){var model=obj(m);if((string)model["job"]!=job)continue;foreach(var s in list(model["steps"])){var spec=obj(s);if((string)spec["name"]==step)return obj(spec["dds"]);}}throw new ArgumentException("Step bindings are missing");}
        string inside(string root,string relative){if(Path.IsPathRooted(relative)||relative.Contains(':')||relative.Replace('\\','/').Split('/').Contains(".."))throw new IOException("Dataset path leaves managed area");string full=Path.GetFullPath(Path.Combine(root,relative)),rel=Path.GetRelativePath(root,full);if(rel==".."||rel.StartsWith(".."+Path.DirectorySeparatorChar))throw new IOException("Dataset path leaves managed area");string at=full;while(at!=root&&at!=null){if((File.Exists(at)||Directory.Exists(at))&&(File.GetAttributes(at)&FileAttributes.ReparsePoint)!=0)throw new IOException("Dataset reparse links are not accepted");at=Path.GetDirectoryName(at);}return full;}
        public RecordFile open(Dictionary<string,object> dd,string mode,int length){var spec=obj(datasets[(string)dd["dsn"]]);if((string)spec["format"]!="fixed"||integer(spec["record_length"])!=length)throw new ArgumentException("Dataset layout differs from source");bool input=(string)spec["role"]=="input";if(input&&mode=="OUTPUT")throw new ArgumentException("Input dataset cannot be overwritten");var f=new RecordFile(inside(input?caseRoot:workRoot,(string)spec["path"]),mode,(string)tables[(string)spec["encoding"]],length);opened.Add(f);return f;}
        public bool relation(object a,string op,object b){int c;if(a is string x&&b is string y){int n=Math.Max(x.Length,y.Length);string table=(string)tables[encoding];c=bytesCompare(encode(pad(x,n),table),encode(pad(y,n),table));}else c=num(a).CompareTo(num(b));return op switch{"="=>c==0,"<>"=>c!=0,">"=>c>0,"<"=>c<0,">="=>c>=0,"<="=>c<=0,_=>throw new ArgumentException("Unknown comparison")};}
        public void copy(Dictionary<string,object> source,Dictionary<string,object> target){int n=integer(obj(datasets[(string)source["dsn"]])["record_length"]);var r=open(source,"INPUT",n);var w=open(target,"OUTPUT",n);string text;while((text=r.read())!=null)w.write(text);r.close();w.close();}
        // OrderBy is used deliberately: List.Sort does not promise stable equal-key order.
        public void sort(Dictionary<string,object> source,Dictionary<string,object> target,int[][] keys){var spec=obj(datasets[(string)source["dsn"]]);int n=integer(spec["record_length"]);var r=open(source,"INPUT",n);var rows=new List<string>();string text;int limit=integer(payload.GetValueOrDefault("max_sort_records",100000));while((text=r.read())!=null){rows.Add(text);if(rows.Count>limit)throw new ArgumentException("Local sort record limit exceeded");}r.close();string table=(string)tables[(string)spec["encoding"]];for(int ki=keys.Length-1;ki>=0;ki--){var key=keys[ki];var cmp=Comparer<string>.Create((a,b)=>{if(key[0]-1+key[1]>a.Length||key[0]-1+key[1]>b.Length)throw new ArgumentException("Sort key extends past record");string x=a.Substring(key[0]-1,key[1]),y=b.Substring(key[0]-1,key[1]);int c;if(key[2]==0)c=bytesCompare(encode(x,table),encode(y,table));else{if(x.Any(v=>v<'0'||v>'9')||y.Any(v=>v<'0'||v>'9'))throw new ArgumentException("Nonnumeric ZD sort key");c=Dec.Parse(x).CompareTo(Dec.Parse(y));}return key[3]==1?-c:c;});rows=rows.OrderBy(x=>x,cmp).ToList();}var w=open(target,"OUTPUT",n);foreach(string row in rows)w.write(row);w.close();}
        public void insert(string table,string[] columns,object[] values)=>db.insert(table,columns,values);
        public void commit()=>db.commit();public void rollback()=>db.rollback();public void finish()=>db.finish();
        public void end_step(){finish();foreach(var f in opened)f.close();opened.Clear();}
        public void Dispose(){try{foreach(var f in opened)f.close();}finally{db.Dispose();}}
    }
}

File: mainframe_migration/migration/targets/templates/java/MfRuntime.java
/* Native Java record and arithmetic behavior shared by generated job files.
 * This is not an interpreter for Python. Each source statement is generated in
 * its job class. Helpers retain fixed-width storage and explicit source rules.
 */
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.math.*;

public final class MfRuntime {
    public static final MathContext MC=new MathContext(80,RoundingMode.HALF_EVEN);
    // Contract objects are checked at the boundary rather than guessed.
    @SuppressWarnings("unchecked") public static Map<String,Object> obj(Object x){if(!(x instanceof Map))throw new IllegalArgumentException("Expected JSON object");return (Map<String,Object>)x;}
    @SuppressWarnings("unchecked") public static List<Object> list(Object x){if(!(x instanceof List))throw new IllegalArgumentException("Expected JSON array");return (List<Object>)x;}
    public static int integer(Object x){return num(x).intValueExact();}
    public static BigDecimal num(Object x){if(x instanceof BigDecimal n)return n;if(x instanceof String||x instanceof Integer||x instanceof Long)return new BigDecimal(x.toString());throw new IllegalArgumentException("Expected exact number");}
    // Perform decimal operations at the explicitly selected arithmetic precision.
    public static Object math(String op,Object a,Object b){BigDecimal x=num(a),y=num(b);return switch(op){case "+"->x.add(y,MC);case "-"->x.subtract(y,MC);case "*"->x.multiply(y,MC);case "/"->x.divide(y,MC);default->throw new IllegalArgumentException("Unknown arithmetic operation");};}
    public static boolean truth(Object x){if(x instanceof Boolean b)return b;if(x instanceof BigDecimal n)return n.signum()!=0;if(x instanceof String s)return !s.isEmpty();throw new IllegalArgumentException("Unsupported condition value");}
    public static byte[] encode(String text,String table){byte[] result=new byte[text.length()];for(int i=0;i<text.length();i++){int n=table.indexOf(text.charAt(i));if(n<0)throw new IllegalArgumentException("Character is not in the source code page");result[i]=(byte)n;}return result;}
    public static String decode(byte[] raw,String table){StringBuilder text=new StringBuilder();for(byte b:raw){int n=b&255;if(n>=table.length())throw new IllegalArgumentException("Byte is not in the source code page");text.append(table.charAt(n));}return text.toString();}
    public static int bytesCompare(byte[] a,byte[] b){for(int i=0;i<Math.min(a.length,b.length);i++){int c=Integer.compare(a[i]&255,b[i]&255);if(c!=0)return c;}return Integer.compare(a.length,b.length);}
    public static String pad(String s,int n){return s.length()>=n?s.substring(0,n):s+" ".repeat(n-s.length());}
    // A reference identifies the caller's live storage, not an eagerly read value.
    public static final class Ref {final Fields owner;final String name;Ref(Fields owner,String name){this.owner=owner;this.name=name;}}
    public static final class Storage {final char[] data;final int offset;Storage(char[] data,int offset){this.data=data;this.offset=offset;}}
    public static final class Fields {
        final Map<String,Map<String,Object>> layout=new LinkedHashMap<>();final Map<String,char[]> buffers=new HashMap<>();final Map<String,Ref> aliases=new HashMap<>();
        // Build group and child fields over the same character buffers.
        public Fields(List<Object> definitions){for(Object x:definitions){var f=obj(x);String name=(String)f.get("name"),root=(String)f.get("root");layout.put(name,f);int size=integer(f.get("offset"))+integer(f.get("length"));char[] old=buffers.get(root);if(old==null||old.length<size){char[] n=new char[size];Arrays.fill(n,' ');if(old!=null)System.arraycopy(old,0,n,0,old.length);buffers.put(root,n);}}
            for(var f:layout.values())if(f.get("kind").equals("number"))set((String)f.get("name"),BigDecimal.ZERO);
            for(var f:layout.values())if(f.containsKey("value"))set((String)f.get("name"),f.get("value"));}
        private Storage storage(String name){var f=layout.get(name);if(f==null)throw new IllegalArgumentException("Unknown source field "+name);String root=(String)f.get("root");int offset=integer(f.get("offset"));if(aliases.containsKey(root)){Ref r=aliases.get(root);Storage s=r.owner.storage(r.name);return new Storage(s.data,s.offset+offset);}return new Storage(buffers.get(root),offset);}
        public int length(String name){return integer(layout.get(name).get("length"));}
        public String raw(String name){Storage s=storage(name);return new String(s.data,s.offset,length(name));}
        public Object get(String name){String text=raw(name);var f=layout.get(name);if(!f.get("kind").equals("number"))return text;if(!text.matches("[0-9]+"))throw new IllegalArgumentException("Nonnumeric source field "+name);return new BigDecimal(text).scaleByPowerOfTen(-integer(f.get("scale")));}
        // Apply receiving-picture truncation and low-order overflow exactly as defined by the supported model.
        public void set(String name,Object value){var f=layout.get(name);int length=length(name);String text;if(f.get("kind").equals("number")){BigInteger number=num(value).abs().scaleByPowerOfTen(integer(f.get("scale"))).toBigInteger();text=number.mod(BigInteger.TEN.pow(length)).toString();text="0".repeat(length-text.length())+text;}else text=pad(value.toString(),length);Storage s=storage(name);text.getChars(0,length,s.data,s.offset);}
        public Ref reference(String name){storage(name);return new Ref(this,name);}
        public void bind(String name,Ref r){var f=layout.get(name);if(!f.get("root").equals(name)||integer(f.get("offset"))!=0||length(name)!=r.owner.length(r.name))throw new IllegalArgumentException("Source CALL linkage differs in width/root");aliases.put(name,r);}
    }
    public static final class RecordFile {
        final InputStream input;final OutputStream output;final String table;final int length;boolean closed=false;
        // Fixed records are binary streams; platform line endings never alter their contents.
        RecordFile(Path path,String mode,String table,int length)throws IOException{this.table=table;this.length=length;if(mode.equals("INPUT")){input=Files.newInputStream(path);output=null;}else if(mode.equals("OUTPUT")){Files.createDirectories(path.getParent());output=Files.newOutputStream(path);input=null;}else throw new IllegalArgumentException("Unsupported OPEN mode");}
        public String read()throws IOException{if(input==null)throw new IllegalStateException("File is not open for input");byte[] raw=input.readNBytes(length);if(raw.length==0)return null;if(raw.length!=length)throw new IOException("Input ends with a partial record");return decode(raw,table);}
        public void write(String text)throws IOException{byte[] raw=encode(text,table);if(output==null||raw.length!=length)throw new IOException("Output record length/mode differs from source");output.write(raw);}
        public void close()throws IOException{if(closed)return;if(input!=null)input.close();if(output!=null)output.close();closed=true;}
    }
    public static final class Context implements AutoCloseable {
        public final Map<String,Object> payload;public final Map<String,Integer> events=new LinkedHashMap<>();public final MfDatabase db;
        final Map<String,Object> datasets,tables;final List<RecordFile> opened=new ArrayList<>();final Path caseRoot,workRoot;final String encoding;
        public Context(Map<String,Object> p)throws Exception{payload=p;datasets=obj(p.get("datasets"));tables=obj(p.get("codepages"));encoding=(String)p.get("collation");caseRoot=Path.of((String)p.get("case_root")).toRealPath();workRoot=Path.of((String)p.get("work_root")).toRealPath();db=new MfDatabase(p);}
        public void hit(String id){events.merge(id,1,Integer::sum);}
        public List<Object> layout(String program){for(Object m:list(payload.get("models"))){var model=obj(m);if(!model.get("job").equals(payload.get("job")))continue;for(Object p:list(model.get("programs"))){var spec=obj(p);if(spec.get("name").equals(program))return list(spec.get("layout"));}}throw new IllegalArgumentException("Program layout is missing");}
        public Map<String,Object> dd(String job,String step){for(Object m:list(payload.get("models"))){var model=obj(m);if(!model.get("job").equals(job))continue;for(Object s:list(model.get("steps"))){var spec=obj(s);if(spec.get("name").equals(step))return obj(spec.get("dds"));}}throw new IllegalArgumentException("Step bindings are missing");}
        private Path inside(Path root,String relative)throws IOException{Path rel=Path.of(relative);if(rel.isAbsolute()||relative.contains(":"))throw new IOException("Dataset path is not relative");for(Path p:rel)if(p.toString().equals(".."))throw new IOException("Dataset path leaves the managed area");Path out=root.resolve(rel).normalize();if(!out.startsWith(root))throw new IOException("Dataset path leaves the managed area");Path check=out;while(check!=null&&!check.equals(root)){if(Files.isSymbolicLink(check))throw new IOException("Dataset symbolic links are not accepted");check=check.getParent();}return out;}
        public RecordFile open(Map<String,Object> dd,String mode,int length)throws Exception{var spec=obj(datasets.get(dd.get("dsn")));if(!spec.get("format").equals("fixed")||integer(spec.get("record_length"))!=length)throw new IllegalArgumentException("Dataset layout differs from source");boolean input=spec.get("role").equals("input");if(input&&mode.equals("OUTPUT"))throw new IllegalArgumentException("Input dataset cannot be overwritten");Path path=inside(input?caseRoot:workRoot,(String)spec.get("path"));RecordFile f=new RecordFile(path,mode,(String)tables.get(spec.get("encoding")),length);opened.add(f);return f;}
        public boolean relation(Object a,String op,Object b){int c;if(a instanceof String&&b instanceof String){String x=(String)a,y=(String)b;int n=Math.max(x.length(),y.length());String table=(String)tables.get(encoding);c=bytesCompare(encode(pad(x,n),table),encode(pad(y,n),table));}else c=num(a).compareTo(num(b));return switch(op){case "="->c==0;case "<>"->c!=0;case ">"->c>0;case "<"->c<0;case ">="->c>=0;case "<="->c<=0;default->throw new IllegalArgumentException("Unknown comparison");};}
        public void copy(Map<String,Object> source,Map<String,Object> target)throws Exception{int n=integer(obj(datasets.get(source.get("dsn"))).get("record_length"));RecordFile r=open(source,"INPUT",n),w=open(target,"OUTPUT",n);String text;while((text=r.read())!=null)w.write(text);r.close();w.close();}
        // List.sort is stable; equal keys retain their source input order, including duplicates.
        public void sort(Map<String,Object> source,Map<String,Object> target,int[][] keys)throws Exception{var spec=obj(datasets.get(source.get("dsn")));int n=integer(spec.get("record_length"));RecordFile r=open(source,"INPUT",n);List<String> rows=new ArrayList<>();String text;int limit=integer(payload.getOrDefault("max_sort_records",100000));while((text=r.read())!=null){rows.add(text);if(rows.size()>limit)throw new IllegalArgumentException("Local sort record limit exceeded");}r.close();String table=(String)tables.get(spec.get("encoding"));for(int ki=keys.length-1;ki>=0;ki--){int[] key=keys[ki];rows.sort((a,b)->{if(key[0]-1+key[1]>a.length()||key[0]-1+key[1]>b.length())throw new IllegalArgumentException("Sort key extends past record");String x=a.substring(key[0]-1,key[0]-1+key[1]),y=b.substring(key[0]-1,key[0]-1+key[1]);int c;if(key[2]==0)c=bytesCompare(encode(x,table),encode(y,table));else{if(!x.matches("[0-9]+")||!y.matches("[0-9]+"))throw new IllegalArgumentException("Nonnumeric ZD sort key");c=new BigDecimal(x).compareTo(new BigDecimal(y));}return key[3]==1?-c:c;});}RecordFile w=open(target,"OUTPUT",n);for(String row:rows)w.write(row);w.close();}
        public void insert(String table,String[] columns,Object[] values)throws Exception{db.insert(table,columns,values);}
        public void commit()throws Exception{db.commit();}
        public void rollback()throws Exception{db.rollback();}
        public void finish()throws Exception{db.finish();}
        public void end_step()throws Exception{finish();for(RecordFile f:opened)f.close();opened.clear();}
        public void close()throws Exception{try{for(RecordFile f:opened)f.close();}finally{db.close();}}
    }
}
